File: src/electron/scope.ts
import path from 'path'
import { realpath } from 'fs/promises'

/**
 * Zero-trust scope enforcement for `fs:*` IPC.
 *
 * The renderer never sends absolute paths over IPC — it sends a `rootId`
 * (a stable identifier registered by main) plus a relative path. This
 * module maintains the rootId → absolute root mapping and verifies every
 * resolved path stays inside its declared root, post symlink resolution.
 *
 * If renderer is compromised, the blast radius is bounded to the union of
 * registered roots — the same files the user could already see by opening
 * a file manager.
 */

const roots = new Map<string, string>()

export function registerRoot(id: string, abs: string): void {
  roots.set(id, path.resolve(abs))
}

export function unregisterRoot(id: string): void {
  roots.delete(id)
}

export function listRoots(): Map<string, string> {
  return new Map(roots)
}

export function getRoot(id: string): string | null {
  return roots.get(id) ?? null
}

/**
 * Resolve `rootId` + `rel` into an absolute path, verifying:
 *   1. `rootId` is a registered root
 *   2. The joined path doesn't escape via `..`
 *   3. After symlink resolution (when the target exists), the realpath
 *      still lies inside the root
 *
 * Throws on any violation. For non-existent targets (writes / probes),
 * returns the joined path without realpath.
 */
export async function resolveScoped(rootId: string, rel: string): Promise<string> {
  const root = roots.get(rootId)
  if (!root) throw new Error(`Root not allowed: ${rootId}`)

  // Normalize the relative path. Strip leading slashes; reject absolute paths.
  if (path.isAbsolute(rel)) throw new Error(`Absolute path not allowed: ${rel}`)
  const safeRel = rel.replace(/^[\\/]+/, '')
  const joined = path.resolve(root, ...safeRel.split(/[\\/]/))

  // Block .. escape pre-realpath.
  if (joined !== root && !joined.startsWith(root + path.sep)) {
    throw new Error(`Path escapes root: ${rel}`)
  }

  // Block symlink escape if the target exists.
  try {
    const real = await realpath(joined)
    if (real !== root && !real.startsWith(root + path.sep)) {
      throw new Error(`Symlink escapes root: ${rel}`)
    }
    return real
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === 'ENOENT') return joined
    throw e
  }
}

/** Reset for tests. */
export function _resetRootsForTesting(): void {
  roots.clear()
}

File: src-tauri/src/scope.rs
// Zero-trust scope enforcement, ported 1:1 from src/electron/scope.ts.
//
// The renderer never sends absolute paths — it sends `(root_id, rel)` and we
// verify every resolved path stays inside the registered root, post symlink
// resolution. Compromised renderer = blast radius bounded to registered roots.

use std::collections::HashMap;
use std::path::{Component, Path, PathBuf};

pub type Roots = HashMap<String, PathBuf>;

pub fn resolve_scoped(roots: &Roots, root_id: &str, rel: &str) -> Result<PathBuf, String> {
    let root = roots
        .get(root_id)
        .ok_or_else(|| format!("Root not allowed: {root_id}"))?;

    let rel_path = Path::new(rel);
    if rel_path.is_absolute() {
        return Err(format!("Absolute path not allowed: {rel}"));
    }

    let mut joined = root.clone();
    for c in rel_path.components() {
        match c {
            Component::Normal(seg) => joined.push(seg),
            Component::CurDir => {}
            // ParentDir / RootDir / Prefix all attempt to escape the scope.
            _ => return Err(format!("Path escapes root: {rel}")),
        }
    }

    // Pre-realpath: defensive normalize check.
    if joined != *root && !joined.starts_with(root) {
        return Err(format!("Path escapes root: {rel}"));
    }

    // Post-realpath: catch symlink escape if the target exists. ENOENT is fine
    // (write to a not-yet-existing file).
    match joined.canonicalize() {
        Ok(real) => {
            let canonical_root = root.canonicalize().unwrap_or_else(|_| root.clone());
            if real != canonical_root && !real.starts_with(&canonical_root) {
                return Err(format!("Symlink escapes root: {rel}"));
            }
            Ok(real)
        }
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(joined),
        Err(e) => Err(format!("canonicalize failed: {e}")),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    fn roots_with(label: &str, dir: &Path) -> Roots {
        let mut r = HashMap::new();
        r.insert(label.into(), dir.to_path_buf());
        r
    }

    #[test]
    fn rejects_unknown_root_id() {
        let roots = HashMap::new();
        let err = resolve_scoped(&roots, "missing", "foo.txt").unwrap_err();
        assert!(err.contains("Root not allowed"), "got: {err}");
    }

    #[test]
    fn rejects_absolute_paths() {
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        let err = resolve_scoped(&roots, "test", "/etc/passwd").unwrap_err();
        assert!(err.contains("Absolute path not allowed"), "got: {err}");
    }

    #[test]
    fn rejects_parent_dir_traversal() {
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        let err = resolve_scoped(&roots, "test", "../outside.txt").unwrap_err();
        assert!(err.contains("Path escapes root"), "got: {err}");
    }

    #[test]
    fn rejects_nested_parent_traversal() {
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        // Even when ../ is mid-path, ParentDir component is rejected outright.
        let err = resolve_scoped(&roots, "test", "papers/../../escape.txt").unwrap_err();
        assert!(err.contains("Path escapes root"), "got: {err}");
    }

    #[test]
    fn allows_simple_relative_paths() {
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        let p = resolve_scoped(&roots, "test", "papers/2024-foo.md").unwrap();
        assert!(p.starts_with(dir.path()));
        assert!(p.ends_with("papers/2024-foo.md"));
    }

    #[test]
    fn collapses_curdir_components() {
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        let p = resolve_scoped(&roots, "test", "./papers/./x.md").unwrap();
        assert!(p.ends_with("papers/x.md"));
    }

    #[test]
    fn allows_writing_to_nonexistent_file() {
        // canonicalize() returns NotFound for paths that don't exist yet.
        // This is the normal case for `fs_write` of a fresh paper.
        let dir = tempfile::tempdir().unwrap();
        let roots = roots_with("test", dir.path());
        let p = resolve_scoped(&roots, "test", "papers/new.md").unwrap();
        assert!(!p.exists());
        assert!(p.starts_with(dir.path()));
    }

    #[cfg(unix)]
    #[test]
    fn rejects_symlink_escape() {
        use std::os::unix::fs::symlink;

        let dir = tempfile::tempdir().unwrap();
        let outside = tempfile::tempdir().unwrap();
        let outside_file = outside.path().join("secret.txt");
        fs::write(&outside_file, "secret").unwrap();

        // Plant a symlink inside the root that points outside.
        symlink(&outside_file, dir.path().join("escape.txt")).unwrap();

        let roots = roots_with("test", dir.path());
        let err = resolve_scoped(&roots, "test", "escape.txt").unwrap_err();
        assert!(err.contains("Symlink escapes root"), "got: {err}");
    }

    #[cfg(unix)]
    #[test]
    fn allows_symlink_within_root() {
        use std::os::unix::fs::symlink;

        let dir = tempfile::tempdir().unwrap();
        let inside_file = dir.path().join("real.md");
        fs::write(&inside_file, "ok").unwrap();
        symlink(&inside_file, dir.path().join("alias.md")).unwrap();

        let roots = roots_with("test", dir.path());
        let p = resolve_scoped(&roots, "test", "alias.md").unwrap();
        assert!(p.starts_with(dir.path()));
    }
}
